### ISSUE TITLE
Support 'lost key' pattern on DB level to satisfy GDPR requirements?

### ISSUE THREAD

**Issue by filimonov** [ClickHouse/ClickHouse#2045]
I know that you, guys, have plans to add some support for UPDATE / DELETE to make it possible to implement 'right to be forgotten' of GDPR.

But also there is another quite nice approach to satisfy GDPR requirements which probably is easier to implement, and allows to process 'forget me' requests very efficiently.

It's called 'lost key' pattern. Idea is simple: all fields containing sensitive (personal) data are encrypted by per-user key, all non-sensitive but important numbers for aggregation can be stored 'as is'. Per user-keys  are stored separately in DB with limited access (in case of ClickHouse it can be just a dictionary). If user decide to be forgotten - it's enough to remove his key from db with key, and no personal information will be available anymore, and at the same time all 'numerical' reports are not affected (you can still see the data of one user and you can see that it was one user, but you can't decrypt it). It will slow down access to fields with sensitive data but in some cases it's not always necessary to decrypt it, and some effecient algorithms can be chosen.

For example can work in a following manner:
```sql
CREATE TABLE xx (
 a Encrypted(String),
 b UInt8,
 c Encrypted(Int8)
 date Date,
 userid UInt8,
) ENGINE ... EncryptionAlgorithm = ... EncryptionKey  = dictGetString( 'keys_storage', 'key', userid);

SELECT decrypt(a) FROM xx; -- decrypted
SELECT a FROM xx; -- raw (encrypted)
 ```

BTW, any update about update / delete progress?

**Comment by filimonov** [comment]
That can also lead to big downgrade of compression ratio (implementing compression before encryption will be not so easy) and together with slowing down data reading / inserting will kill all the advantages.

**Comment by alexey-milovidov** [comment]
This is very clever idea!

Probably we can start with encryption/decryption functions.
Special encrypted data types and syntax in table definition may go later.

Compression rate will suffer. But not so much if we just encrypt only few certain fields out of many.
CPU cost will be not so much if we will be able to utilize AES-NI instructions.

If we encrypt values independently (one by one) we should also add some unused random bits before encryption, to avoid correlated results.

I'm not an expert in crypto. Could you please propose some details about possible interfaces of that functions?

**Comment by filimonov** [comment]
To utilize AES-NI instruction the only algorithm of choice is AES (in one of the modes).

I think that CTR mode or AES should be nice - reasonable good, fast, work in stream mode (implementations should support SIMD, probably should support parallelization).

For algorithm itself you need to provide initialization vector (IV), the plaintext (or ciphertext to decode) and key (one of the sizes: 128, 192, 256 bits). 

You can use the constant IV: that will simplify some types of attacks, by giving property: 
```if encrypt(A,KEY) = encrypt(B,KEY) then A=B```
If the key will be used / removed per user that's doesn't sound like an issue, and can even be desirable property for some kinds of encrypted data processing (so you will see that some encrypted attribute of user X remains fixed in time).

Key and initialization vector can be generated from passphrase string and salt using some key derivation function (for example PBKDF2 from PKCS#5). But that will require extra initialization steps before encrypt/decrypt, so for ClickHouse it's rather better to have key/IV 'ready to use' - i.e. binary and with proper length (16/24/32 bytes depending of chosen option), without extra step of generation key/IV from passphrase/salt.

You can play a bit with encrypt/decrypt using [openssl cli](https://wiki.openssl.org/index.php/Enc):
```bash
➜ echo 'aaaaa' | openssl enc -K 00000000000000000000000000000000 -iv 00000000000000000000000000000000 -aes-128-ctr -base64
B4gqtY6A
➜ echo 'B4gqtY6A' | openssl enc -K 00000000000000000000000000000000 -iv 00000000000000000000000000000000 -aes-128-ctr -base64 -d
aaaaa
# Openssl also has builtin benchmarks that can by run from bash like that:
➜ openssl speed
```

About interface: key and text are required. IV - may be can be provided as optional param? Text is any binary data, key is binary data of configured length, result is binary data.

Also (may be?) it will be better to leave possibility to use other algorithms in future, something like 
`encrypt('aes-128-ctr')(data, key)` ?

In C++ you can use one of avaliable [cryptography libraries](https://en.wikipedia.org/wiki/Comparison_of_cryptography_libraries). It looks like most of them should AES-NI instruction. 

Benchmark of one of them give some feeling of algorithm speed:
https://www.cryptopp.com/benchmarks.html 

> I'm not an expert in crypto.

I didn't have practical experience with crypto in last 10 years, so just consider written above as IMHO.

**Comment by filimonov** [comment]
Update: as Clickhouse stores column data in blocks and each block is compressed and can be processed separately, encryprion can also be done on block level. In that case initialization vector could be unique per block.

**Comment by jenyacazacu** [comment]
is there any update on this, any plans to release encryption at rest in a future version?

**Comment by jondeandres** [comment]
Any update on this? It's in the roadmap for next months?

**Comment by alexey-milovidov** [comment]
This task is assigned to @bufuchangfeng (Yuchen Dong).

**Comment by shsab** [comment]
Any update on encryption?

**Comment by jondeandres** [comment]
@bufuchangfeng any update on this? Thanks

**Comment by alexey-milovidov** [comment]
Implemented in #11844.
